monitoring/uss_qualifier/scenarios/scenario.py

```diff
@@ -22,17 +33,247 @@ class TestScenario(ABC):
     """
 
     documentation: TestScenarioDocumentation
+    on_failed_check: Optional[Callable[[FailedCheck], None]] = None
+    _scenario_report: Optional[TestScenarioReport] = None
+    _current_case: Optional[TestCaseDocumentation] = None
+    _case_report: Optional[TestCaseReport] = None
+    _current_step: Optional[TestStepDocumentation] = None
+    _step_report: Optional[TestStepReport] = None
 
     def __init__(self):
         self.documentation = parse_documentation(self.__class__)
 
     @abstractmethod
-    # TODO: have `run` interact with an encapsulated portion of a report, or return contributions to a report
     def run(self):
         raise NotImplementedError(
             "A concrete test scenario must implement `run` method"
         )
 
+    def me(self) -> str:
+        return inspection.fullname(self.__class__)
+
+    def _make_scenario_report(self) -> None:
+        self._scenario_report = TestScenarioReport(
+            name=self.documentation.name,
+            documentation_url=self.documentation.url,
+            start_time=StringBasedDateTime(datetime.utcnow()),
+            cases=[],
+        )
+
+    def begin_test_scenario(self) -> None:
+        if self._scenario_report is not None:
+            raise RuntimeError(
+                f"Test scenario `{self.me()}` had already begun when begin_test_scenario was called"
+            )
+        self._make_scenario_report()
+
+    def begin_test_case(self, name: str) -> None:
+        if self._scenario_report is None:
```

Comment:
nit: Those checks may perhaps be extracted to a dedicated method / decorator to limit repetition and improve readability.

Reply:
Yeah, I'm open to making that kind of change (perhaps in a different PR) -- the challenge is that the error messages are mostly slightly different from each other and I think good, precise error messages are important